### PR TITLE
Simplify `defn.FunctionOf.unapply`

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -1990,10 +1990,7 @@ trait Applications extends Compatibility {
         // the arity of that function, otherise -1.
         def paramCount(ref: TermRef) =
           val formals = ref.widen.firstParamTypes
-          if formals.length > idx then
-            formals(idx) match
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala:1994]: An alternative fix for this case is
```diff
-            formals(idx) match
+            formals(idx).dealias match
```

File: compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala [hunk 2]
```diff
@@ -408,10 +408,9 @@ class CheckCaptures extends Recheck, SymTransformer:
       else if meth == defn.Caps_unsafeUnbox then
         mapArgUsing(_.forceBoxStatus(false))
       else if meth == defn.Caps_unsafeBoxFunArg then
-        mapArgUsing:
-          case defn.FunctionOf(paramtpe :: Nil, restpe, isContextual) =>
-            defn.FunctionOf(paramtpe.forceBoxStatus(true) :: Nil, restpe, isContextual)
-
+        mapArgUsing: tp =>
+          val defn.FunctionOf(paramtpe :: Nil, restpe, isContextual) = tp.strippedDealias: @unchecked
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala:412]: In the previous and current implementations, we need to strip the capture set from `tp` to get to the underlying function type. Is this intended, or are we losing part of the capture set that the resulting function needs?

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/cc/CheckCaptures.scala:412]: I also noticed that previously we stripped the capture set in `FunctionOf` for `FunctionN` but not for `PolyFunction`. This inconsistency could have been noticed on functions with erased parameters. Now we do not strip capture sets in any of the cases in `FunctionOf`.

File: compiler/src/dotty/tools/dotc/core/Definitions.scala [hunk 3]
```diff
@@ -1116,16 +1116,13 @@ class Definitions {
       else
         FunctionType(args.length, isContextual).appliedTo(args ::: resultType :: Nil)
     def unapply(ft: Type)(using Context): Option[(List[Type], Type, Boolean)] = {
-      ft.dealias match
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/core/Definitions.scala:1119]: What is the reason to drop the dealias in the extractor?

**Review by odersky — COMMENTED** [review]
Otherwise LGTM

File: compiler/src/dotty/tools/dotc/core/Definitions.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1119]: There are a few reasons:
* There are few places where that type is not already dealiased
* In a few cases we used to repeatedly dealias the type in pattern match extractors for functions. Not dealiassing here forces the use site to dealias a single time. Sometimes this happens in cases of a single pattern match and sometimes in the recursion of a `TypeMap`.
* This aligns the implementation closer to `RefinedFunctionOf`, `PolyFunctionOf` and `FunctionTypeOfMethod` (in #18443).
* It brings us one step closer to being able to have a `derivedFunctionTypeOf` method.

Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/Definitions.scala:1119]: OK, I agree it's good to align the unapply's towards no dealising.

**Review by odersky — APPROVED** [review]

**Pull request merged**